models/hashflow/ethereum/hashflow_ethereum_trades.sql

```diff
@@ -0,0 +1,407 @@
+        and `to` in ('0x455a3B3Be6e7C8843f2b03A1cA22A5a5727ef5C4','0x043389f397ad72619d05946f5f35426a7ace6613')
+        and substring(input, 1, 4) in ('0xd0529c02',  -- token to eth
+                                       '0xe43d9733') -- eth to token
+        and p.symbol = 'WETH'
+),
+
+new_pool as (
+    -- subquery for including new pools created on 2022-04-09
+    -- same trade event abi, effectively only from table hashflow.pool_evt_trade since 2022-04-09
+    select
+        l.evt_index as composite_index,
+        null as source, -- no join on call for this batch, refer to metabase for source info
+        tx.block_time as block_time,
+        tx.hash as tx_hash,
+        true as fill_status, -- without call we are only logging successful fills
+        null as method_id, -- without call we don't have function call info
+        tx.to as router_contract, -- taking top level contract called in tx as router, not necessarily HF contract
+        l.pool as pool,
+        tx.from as trader,
+        l.`quoteToken` as maker_token,
+        l.`baseToken` as taker_token,
+        case when l.`quoteToken` = '0x0000000000000000000000000000000000000000' then 'ETH'
+            else mp.symbol end as maker_symbol,
+        case when l.`baseToken` = '0x0000000000000000000000000000000000000000' then 'ETH'
+            else tp.symbol end as taker_symbol,
+        l.`quoteTokenAmount`/power(10, mp.decimals) as maker_token_amount,
+        l.`baseTokenAmount`/power(10, tp.decimals) as taker_token_amount,
+        coalesce(
+                l.`baseTokenAmount`/power(10, tp.decimals) * tp.price,
+                l.`quoteTokenAmount`/power(10, mp.decimals) * mp.price) as amount_usd
+    from hashflow_pool_evt_trade l
+    join ethereum_transactions tx on tx.hash = l.evt_tx_hash
+    left join prices_usd tp on tp.minute = date_trunc('minute', tx.block_time)
+        and tp.contract_address =
+            case when l.`baseToken` = '0x0000000000000000000000000000000000000000'
+                then '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2' else l.`baseToken` end
+    left join prices_usd mp on mp.minute = date_trunc('minute', tx.block_time)
+        and mp.contract_address =
+            case when l.`quoteToken` = '0x0000000000000000000000000000000000000000'
+                then '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2' else l.`quoteToken` end
+    WHERE l.evt_block_time > '2022-04-08' -- necessary filter to only include new trades
+),
+
+all_trades as (
+    select
+        -1 as composite_index,
+        -- was decoding from trace, no log_index, only single swap exist so works as PK
+        '0x00' as source,
+        -- all from native front end, no integration yet
+        *
+    from legacy_routers
+
+    union all
+
+    select * from new_pool
+
+    {% if is_incremental() %}
+    {% else %}
+
+    union all
+
+    select * from legacy_router_w_integration
+
+    union all
+
+    select * from new_router
+
+    {% endif %}
```

Comment:
Since `new_router` and `new_pool` CTEs have overlapping transaction events in the source, 

- we can add another UNION CTE that unions `new_router` and `new_pool`
- and group by the unique keys `['block_date', 'composite_index', 'tx_hash']` 
- (with `coalesce(collect_list(...))` or `first_value(..., true)` on columns like `['method_id', 'pool', 'source', ...]`) 

before the all_trades UNION.

Also, in this complex model, the incremental process may need to be handle with care.

Reply:
that makes sense - will update to remove the overlapping elements this way

r.e. your incremental comment -> did you have specific parts of my current approach that you were referring to?

Reply:
Only `new_router` this part, as I see it only runs when full refresh. So if we do the CTE union before `all_trades` with `new_pool`, the incremental logic also need to be migrated to that CTE.